azurelinuxagent/ga/periodic_operation.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+
+import datetime
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.future import ustr
+
+
+class PeriodicOperation(object):
+    '''
+    Instances of PeriodicOperation are tasks that are executed only after the given
+    time period has elapsed.
+    '''
+
+    # To prevent flooding the log with error messages we report failures at most every hour
+    _LOG_WARNING_PERIOD = datetime.timedelta(minutes=60)
+
+    def __init__(self, name, operation, period):
+        self._name = name
+        self._operation = operation
+        self._period = period
+        self._last_run = None
+        self._last_log_warning = None
+
+    def run(self):
+        try:
+            if self._last_run is None or datetime.datetime.utcnow() >= self._last_run + self._period:
+                try:
+                    self._operation()
+                finally:
+                    self._last_run = datetime.datetime.utcnow()
+        except Exception as e:
+            if self._last_log_warning is None or datetime.datetime.utcnow() >= self._last_log_warning + self._LOG_WARNING_PERIOD:
+                logger.warn("[PERIODIC] Failed to {0}: {1}", self._name, ustr(e))
```

Comment:
I am a bit concerned about how the Periodic logger is set up here. If the _operation function fails for multiple reasons within the same hour and it does not have its own try-except, we might miss out on the errors as this implementation logs errors based on the operation rather than the type of the error (as we do it in the PeriodicLogger). 

I feel instead of setting a periodic logger, we should just log at the time frequency that the operation is called in so that we dont miss out on key errors. Or maybe have a cache per error maybe?

Reply:
(addresed in the reply to your first comment)